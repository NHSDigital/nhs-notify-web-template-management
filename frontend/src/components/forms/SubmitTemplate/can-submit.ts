import { TemplateDto } from 'nhs-notify-backend-client';
import { LetterTemplate } from 'nhs-notify-web-template-management-utils';

function canSubmitLetterTemplate(template: LetterTemplate): boolean {
  const { pdfTemplate, testDataCsv, proofs } = template.files;

  const virusScanStatuses = [
    pdfTemplate.virusScanStatus,
    testDataCsv?.virusScanStatus,
    ...(proofs?.map((proof) => proof.virusScanStatus) ?? []),
  ];

  return !virusScanStatuses.some((status) => status !== 'PASSED');
}

export function canSubmit(template: TemplateDto): boolean {
  if (!template) {
    return false;
  }

  if (template.templateStatus !== 'NOT_YET_SUBMITTED') {
    return false;
  }

  if (template.templateType === 'LETTER') {
    return canSubmitLetterTemplate(template as LetterTemplate);
  }

  return true;
}
