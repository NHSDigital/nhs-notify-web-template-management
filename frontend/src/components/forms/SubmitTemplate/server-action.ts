'use server';

import { redirect, RedirectType } from 'next/navigation';
import { getTemplate, saveTemplate } from '@utils/form-actions';
import { z } from 'zod';
import { validateChannelTemplate } from '@utils/validate-template';
import { TemplateStatus } from 'nhs-notify-web-template-management-utils';
import { logger } from 'nhs-notify-web-template-management-utils/logger';

const $TemplateIdSchema = z.string();

export async function submitTemplate(route: string, formData: FormData) {
  const { success, data: templateId } = $TemplateIdSchema.safeParse(
    formData.get('templateId')
  );

  if (!success) {
    return redirect('/invalid-template', RedirectType.replace);
  }

  const template = await getTemplate(templateId);

  const validatedTemplate = validateChannelTemplate(template);

  if (!validatedTemplate) {
    return redirect('/invalid-template', RedirectType.replace);
  }

  try {
    await saveTemplate({
      ...validatedTemplate,
      templateStatus: TemplateStatus.SUBMITTED,
    });
  } catch (error) {
    logger.error('Failed to submit template', {
      error,
      templateId,
    });

    throw error;
  }

  return redirect(`/${route}/${templateId}`, RedirectType.push);
}
