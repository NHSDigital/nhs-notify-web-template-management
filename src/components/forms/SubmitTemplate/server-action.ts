'use server';

import { redirect, RedirectType } from 'next/navigation';
import { getSession, saveTemplate, sendEmail } from '@utils/form-actions';
import { createTemplateFromSession, validateTemplate } from '@domain/templates';
import { logger } from '@utils/logger';
import { z } from 'zod';

const $SessionIdSchema = z.string();

export async function submitTemplate(route: string, formData: FormData) {
  const { success, data: sessionId } = $SessionIdSchema.safeParse(
    formData.get('sessionId')
  );

  if (!success) {
    return redirect('/invalid-session', RedirectType.replace);
  }

  const session = await getSession(sessionId);

  if (!session) {
    return redirect('/invalid-session', RedirectType.replace);
  }

  try {
    const templateDTO = createTemplateFromSession(session);

    const validatedTemplate = validateTemplate(templateDTO);

    const templateEntity = await saveTemplate(validatedTemplate);

    await sendEmail(
      templateEntity.id,
      templateEntity.name,
      templateEntity.fields!.content
    );

    return redirect(`/${route}/${templateEntity.id}`, RedirectType.push);
  } catch (error) {
    logger.error('Failed to submit template', {
      error,
      sessionId,
    });

    throw error;
  }
}
