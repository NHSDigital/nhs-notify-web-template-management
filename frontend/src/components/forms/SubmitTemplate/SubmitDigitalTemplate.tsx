'use client';

import { FC, useActionState } from 'react';
import { WarningCallout } from 'nhsuk-react-components';
import { SubmitTemplatePageComponentProps } from 'nhs-notify-web-template-management-utils';
import content from '@content/content';
import { NHSNotifyFormWrapper } from '@molecules/NHSNotifyFormWrapper/NHSNotifyFormWrapper';
import { getBasePath } from '@utils/get-base-path';
import { submitTemplate } from '@forms/SubmitTemplate/server-action';
import { NHSNotifyMain } from '@atoms/NHSNotifyMain/NHSNotifyMain';
import { NHSNotifyButton } from '@atoms/NHSNotifyButton/NHSNotifyButton';

export const SubmitDigitalTemplate: FC<SubmitTemplatePageComponentProps> = ({
  templateName,
  templateId,
  goBackPath,
  submitPath,
}) => {
  const {
    pageHeading,
    warningCalloutLabel,
    warningCalloutText,
    submitChecklistHeading,
    submitChecklistIntroduction,
    submitChecklistItems,
    submitChecklistParagraphs,
    goBackButtonText,
    buttonText,
  } = content.components.submitTemplate;

  const [_, action] = useActionState(submitTemplate, submitPath);

  return (
    <NHSNotifyMain>
      <div className='nhsuk-grid-row'>
        <div className='nhsuk-grid-column-two-thirds'>
          <h1>
            {pageHeading} {`'${templateName}'`}
          </h1>
          <WarningCallout>
            <WarningCallout.Label headingLevel='h2'>
              {warningCalloutLabel}
            </WarningCallout.Label>
            <p>{warningCalloutText}</p>
          </WarningCallout>
          <h2 className='nhsuk-heading-m'>{submitChecklistHeading}</h2>
          <p>{submitChecklistIntroduction}</p>
          <ul>
            {submitChecklistItems.map((item) => (
              <li key={`submit-list-${item.slice(0, 5)}`}>{item}</li>
            ))}
          </ul>
          {submitChecklistParagraphs.map((item) => (
            <p key={`submit-paragraph-${item.slice(0, 5)}`}>{item}</p>
          ))}
          <NHSNotifyFormWrapper formId='submit-template-form' action={action}>
            <input
              type='hidden'
              name='templateId'
              value={templateId}
              readOnly
            />
            <NHSNotifyButton
              secondary
              id='go-back-button'
              className='nhsuk-u-margin-right-3'
              href={`${getBasePath()}/${goBackPath}/${templateId}`}
            >
              {goBackButtonText}
            </NHSNotifyButton>
            <NHSNotifyButton id='submit-template-button'>
              {buttonText}
            </NHSNotifyButton>
          </NHSNotifyFormWrapper>
        </div>
      </div>
    </NHSNotifyMain>
  );
};
