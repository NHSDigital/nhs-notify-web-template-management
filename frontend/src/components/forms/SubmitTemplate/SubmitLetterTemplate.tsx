'use client';

import { FC, useActionState } from 'react';
import { WarningCallout } from 'nhsuk-react-components';
import content from '@content/content';
import { NHSNotifyFormWrapper } from '@molecules/NHSNotifyFormWrapper/NHSNotifyFormWrapper';
import { getBasePath } from '@utils/get-base-path';
import { submitTemplate } from '@forms/SubmitTemplate/server-action';
import { NHSNotifyMain } from '@atoms/NHSNotifyMain/NHSNotifyMain';
import { NHSNotifyButton } from '@atoms/NHSNotifyButton/NHSNotifyButton';

// breaking sonarqube rules about duplication, but the code is temporary
// remove this in https://nhsd-jira.digital.nhs.uk/browse/CCM-9752
// BEGIN-NOSCAN
export const SubmitLetterTemplateProofingDisabled: FC<{
  templateName: string;
  templateId: string;
  action: (payload: FormData) => void;
}> = ({ templateName, templateId, action }) => {
  const {
    proofingFlagDisabled: {
      afterSubmissionHeading,
      afterSubmissionText,
      buttonText,
      goBackButtonText,
      goBackPath,
      pageHeading,
      submitChecklistHeading,
      submitChecklistIntroduction,
      submitChecklistItems,
      warningCalloutChecklistIntroduction,
      warningCalloutChecklistItems,
      warningCalloutLabel,
    },
  } = content.components.submitLetterTemplate;

  return (
    <NHSNotifyMain>
      <div className='nhsuk-grid-row'>
        <div className='nhsuk-grid-column-two-thirds'>
          <h1>
            {pageHeading} {`'${templateName}'`}
          </h1>
          <h2 className='nhsuk-heading-s'>{submitChecklistHeading}</h2>
          <p>{submitChecklistIntroduction}</p>
          <ul>
            {submitChecklistItems.map((item) => (
              <li key={`submit-list-${item.slice(0, 5)}`}>{item}</li>
            ))}
          </ul>
          <h2 className='nhsuk-heading-s'>{afterSubmissionHeading}</h2>
          {afterSubmissionText.map((item) => (
            <p key={`after-submission-paragraph-${item.slice(0, 5)}`}>{item}</p>
          ))}
          <WarningCallout>
            <WarningCallout.Label headingLevel='h2'>
              {warningCalloutLabel}
            </WarningCallout.Label>
            <p>{warningCalloutChecklistIntroduction}</p>
            <ul>
              {warningCalloutChecklistItems.map((item) => (
                <li key={`warning-callout-list-${item.slice(0, 5)}`}>{item}</li>
              ))}
            </ul>
          </WarningCallout>
          <NHSNotifyFormWrapper formId='submit-template-form' action={action}>
            <input
              type='hidden'
              name='templateId'
              value={templateId}
              readOnly
            />
            <NHSNotifyButton
              secondary
              id='go-back-button'
              className='nhsuk-u-margin-right-3'
              href={`${getBasePath()}/${goBackPath}/${templateId}`}
            >
              {goBackButtonText}
            </NHSNotifyButton>
            <NHSNotifyButton id='submit-template-button'>
              {buttonText}
            </NHSNotifyButton>
          </NHSNotifyFormWrapper>
        </div>
      </div>
    </NHSNotifyMain>
  );
};

export const SubmitLetterTemplate: FC<{
  templateName: string;
  templateId: string;
}> = ({ templateName, templateId }) => {
  const {
    buttonText,
    goBackButtonText,
    goBackPath,
    intro,
    pageHeading,
    submitChecklistHeading,
    submitChecklistIntroduction,
    submitChecklistItems,
    warningCalloutChecklistIntroduction,
    warningCalloutChecklistItems,
    warningCalloutLabel,
  } = content.components.submitLetterTemplate;

  const [_, action] = useActionState(submitTemplate, 'LETTER');

  if (process.env.NEXT_PUBLIC_ENABLE_PROOFING !== 'true') {
    return (
      <SubmitLetterTemplateProofingDisabled
        templateName={templateName}
        templateId={templateId}
        action={action}
      />
    );
  }

  return (
    <NHSNotifyMain>
      <div className='nhsuk-grid-row'>
        <div className='nhsuk-grid-column-two-thirds'>
          <h1>
            {pageHeading} {`'${templateName}'`}
          </h1>
          <p className='nhsuk-body-l'>{intro}</p>
          <h2 className='nhsuk-heading-s'>{submitChecklistHeading}</h2>
          <p>{submitChecklistIntroduction}</p>
          <ul>
            {submitChecklistItems.map((item) => (
              <li key={`submit-list-${item.slice(0, 5)}`}>{item}</li>
            ))}
          </ul>
          <WarningCallout>
            <WarningCallout.Label headingLevel='h2'>
              {warningCalloutLabel}
            </WarningCallout.Label>
            <p>{warningCalloutChecklistIntroduction}</p>
            <ul>
              {warningCalloutChecklistItems.map((item) => (
                <li key={`warning-callout-list-${item.slice(0, 5)}`}>{item}</li>
              ))}
            </ul>
          </WarningCallout>
          <NHSNotifyFormWrapper formId='submit-template-form' action={action}>
            <input
              type='hidden'
              name='templateId'
              value={templateId}
              readOnly
            />
            <NHSNotifyButton
              secondary
              id='go-back-button'
              className='nhsuk-u-margin-right-3'
              href={`${getBasePath()}/${goBackPath}/${templateId}`}
            >
              {goBackButtonText}
            </NHSNotifyButton>
            <NHSNotifyButton id='submit-template-button'>
              {buttonText}
            </NHSNotifyButton>
          </NHSNotifyFormWrapper>
        </div>
      </div>
    </NHSNotifyMain>
  );
};
// END-NOSCAN
