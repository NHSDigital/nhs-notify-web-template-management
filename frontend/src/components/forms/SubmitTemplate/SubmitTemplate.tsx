'use client';

import { FC } from 'react';
import { WarningCallout, Button } from 'nhsuk-react-components';
import { SubmitTemplatePageComponentProps } from 'nhs-notify-web-template-management-utils';
import { submitTemplateContent } from '@content/content';
import { NHSNotifyFormWrapper } from '@molecules/NHSNotifyFormWrapper/NHSNotifyFormWrapper';
import { getBasePath } from '@utils/get-base-path';
import { submitTemplate } from '@forms/SubmitTemplate/server-action';
import { NHSNotifyMain } from '@atoms/NHSNotifyMain/NHSNotifyMain';

export const SubmitTemplate: FC<SubmitTemplatePageComponentProps> = ({
  templateName,
  templateId,
  goBackPath,
  submitPath,
}) => {
  const {
    pageHeading,
    warningCalloutLabel,
    warningCalloutText,
    submitChecklistHeading,
    submitChecklistIntroduction,
    submitChecklistItems,
    submitChecklistParagraphs,
    goBackButtonText,
    buttonText,
  } = submitTemplateContent;

  return (
    <NHSNotifyMain>
      <div className='nhsuk-grid-row'>
        <div className='nhsuk-grid-column-two-thirds'>
          <h1>
            {pageHeading} {`'${templateName}'`}
          </h1>
          <WarningCallout>
            <WarningCallout.Label headingLevel='h2'>
              {warningCalloutLabel}
            </WarningCallout.Label>
            <p>{warningCalloutText}</p>
          </WarningCallout>
          <h2 className='nhsuk-heading-m'>{submitChecklistHeading}</h2>
          <p>{submitChecklistIntroduction}</p>
          <ul>
            {submitChecklistItems.map((item) => (
              <li key={`submit-list-${item.slice(0, 5)}`}>{item}</li>
            ))}
          </ul>
          {submitChecklistParagraphs.map((item) => (
            <p key={`submit-paragraph-${item.slice(0, 5)}`}>{item}</p>
          ))}
          <NHSNotifyFormWrapper
            formId='submit-template-form'
            action={submitTemplate.bind(null, submitPath)}
          >
            <input
              type='hidden'
              name='templateId'
              value={templateId}
              readOnly
            />
            <Button
              preventDoubleClick
              secondary
              id='go-back-button'
              className='nhsuk-u-margin-right-3'
              href={`${getBasePath()}/${goBackPath}/${templateId}`}
            >
              {goBackButtonText}
            </Button>
            <Button preventDoubleClick id='submit-template-button'>{buttonText}</Button>
          </NHSNotifyFormWrapper>
        </div>
      </div>
    </NHSNotifyMain>
  );
};
