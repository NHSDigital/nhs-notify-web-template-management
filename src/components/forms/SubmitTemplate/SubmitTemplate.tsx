'use client';

import { FC } from 'react';
import { WarningCallout, Button, BackLink } from 'nhsuk-react-components';
import { SubmitTemplatePageComponentProps } from '@utils/types';
import { submitNhsAppTemplateContent } from '@content/content';
import { NHSNotifyFormWrapper } from '@molecules/NHSNotifyFormWrapper/NHSNotifyFormWrapper';
import { getBasePath } from '@utils/get-base-path';
import { submitTemplate } from './server-action';

export const SubmitTemplate: FC<SubmitTemplatePageComponentProps> = ({
  templateName,
  sessionId,
  goBackPath,
}) => {
  const {
    backLinkText,
    pageHeading,
    warningCalloutLabel,
    warningCalloutText,
    submitChecklistHeading,
    submitChecklistIntroduction,
    submitChecklistItems,
    submitChecklistParagraphs,
    buttonText,
  } = submitNhsAppTemplateContent;

  return (
    <div className='nhsuk-grid-row'>
      <BackLink
        href={`${getBasePath()}/${goBackPath}/${sessionId}`}
        className='nhsuk-u-margin-bottom-7 nhsuk-u-margin-left-3'
      >
        {backLinkText}
      </BackLink>
      <div className='nhsuk-grid-column-two-thirds'>
        <h1>
          {pageHeading} {`'${templateName}'`}
        </h1>
        <WarningCallout>
          <WarningCallout.Label headingLevel='h2'>
            {warningCalloutLabel}
          </WarningCallout.Label>
          <p>{warningCalloutText}</p>
        </WarningCallout>
        <h2 className='nhsuk-heading-m'>{submitChecklistHeading}</h2>
        <p>{submitChecklistIntroduction}</p>
        <ul>
          {submitChecklistItems.map((item) => (
            <li key={`submit-list-${item.slice(0, 5)}`}>{item}</li>
          ))}
        </ul>
        {submitChecklistParagraphs.map((item) => (
          <p key={`submit-paragraph-${item.slice(0, 5)}`}>{item}</p>
        ))}
        <NHSNotifyFormWrapper
          formId='submit-template-form'
          action={submitTemplate}
        >
          <input type='hidden' name='sessionId' value={sessionId} readOnly />
          <Button id='submit-template-button'>{buttonText}</Button>
        </NHSNotifyFormWrapper>
      </div>
    </div>
  );
};
