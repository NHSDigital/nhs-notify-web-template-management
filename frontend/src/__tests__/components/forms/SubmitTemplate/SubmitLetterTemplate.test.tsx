import { SubmitLetterTemplate } from '@forms/SubmitTemplate/SubmitLetterTemplate';
import { render } from '@testing-library/react';

jest.mock('@utils/amplify-utils');

jest.mock('react', () => {
  const originalModule = jest.requireActual('react');

  return {
    ...originalModule,
    useActionState: jest.fn((action, initialState) => {
      return [initialState, action];
    }),
  };
});

jest.mock('@forms/SubmitTemplate/server-action', () => ({
  submitTemplate: '/action',
}));

describe('SubmitLetterTemplate component', () => {
  it('should render', () => {
    const container = render(
      <SubmitLetterTemplate
        templateId='template-id'
        templateName='template-name'
        goBackPath='example'
        submitPath='example-submit'
      />
    );

    expect(container.asFragment()).toMatchSnapshot();
  });
});
