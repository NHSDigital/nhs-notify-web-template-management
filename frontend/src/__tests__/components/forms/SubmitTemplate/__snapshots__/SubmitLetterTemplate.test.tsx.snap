// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmitLetterTemplate component should render 1`] = `
<DocumentFragment>
  <main
    class="nhsuk-main-wrapper"
    id="maincontent"
    role="main"
  >
    <div
      class="nhsuk-grid-row"
    >
      <div
        class="nhsuk-grid-column-two-thirds"
      >
        <h1>
          Approve and submit 'template-name'
        </h1>
        <p
          class="nhsuk-body-l"
        >
          When you submit a letter template, it will be used by NHS Notify to set up the messages you want to send.
        </p>
        <h2
          class="nhsuk-heading-s"
        >
          Before you submit this template
        </h2>
        <p>
          Check that your template proof:
        </p>
        <ul
          class="nhsuk-list nhsuk-list--bullet"
        >
          <li>
            looks exactly as you expect your recipient to get it
          </li>
          <li>
            uses personalisation as you expect
          </li>
          <li>
            shows QR codes correctly (if used)
          </li>
        </ul>
        <div
          class="nhsuk-warning-callout"
        >
          <h2
            class="nhsuk-warning-callout__label"
          >
            <span
              role="text"
            >
              <span
                class="nhsuk-u-visually-hidden"
              >
                Important: 
              </span>
              Important
            </span>
          </h2>
          <p>
            You cannot edit a template after you've approved and submitted it. You can only replace it with a new template.
          </p>
        </div>
        <form
          action="/action"
        >
          <input
            name="form-id"
            readonly=""
            type="hidden"
            value="submit-template-form"
          />
          <input
            name="csrf_token"
            readonly=""
            type="hidden"
            value="no_token"
          />
          <input
            name="templateId"
            readonly=""
            type="hidden"
            value="template-id"
          />
          <a
            aria-disabled="false"
            class="nhsuk-button nhsuk-button--secondary nhsuk-u-margin-right-3"
            draggable="false"
            href="/templates/preview-letter-template/template-id"
            id="go-back-button"
            role="button"
          >
            Go back
          </a>
          <button
            aria-disabled="false"
            class="nhsuk-button"
            id="submit-template-button"
            type="submit"
          >
            Approve and submit
          </button>
        </form>
      </div>
    </div>
  </main>
</DocumentFragment>
`;

exports[`SubmitLetterTemplate component should render with client proofing disabled 1`] = `
<DocumentFragment>
  <main
    class="nhsuk-main-wrapper"
    id="maincontent"
    role="main"
  >
    <div
      class="nhsuk-grid-row"
    >
      <div
        class="nhsuk-grid-column-two-thirds"
      >
        <h1>
          Submit 'template-name'
        </h1>
        <h2
          class="nhsuk-heading-s"
        >
          Before you submit
        </h2>
        <p>
          You should check that your template:
        </p>
        <ul
          class="nhsuk-list nhsuk-list--bullet"
        >
          <li>
            is approved by the relevant stakeholders in your team
          </li>
          <li>
            does not have any spelling errors
          </li>
          <li>
            is formatted correctly
          </li>
        </ul>
        <h2
          class="nhsuk-heading-s"
        >
          After you submit this template
        </h2>
        <p>
          Our service team will send you a proof of this letter template by email.
        </p>
        <p>
          This email will also tell you what you need to do next.
        </p>
        <div
          class="nhsuk-warning-callout"
        >
          <h2
            class="nhsuk-warning-callout__label"
          >
            <span
              role="text"
            >
              <span
                class="nhsuk-u-visually-hidden"
              >
                Important: 
              </span>
              Important
            </span>
          </h2>
          <p>
            You cannot edit a template after you've submitted it. You can only replace it with a new template.
          </p>
        </div>
        <form
          action="/action"
        >
          <input
            name="form-id"
            readonly=""
            type="hidden"
            value="submit-template-form"
          />
          <input
            name="csrf_token"
            readonly=""
            type="hidden"
            value="no_token"
          />
          <input
            name="templateId"
            readonly=""
            type="hidden"
            value="template-id"
          />
          <a
            aria-disabled="false"
            class="nhsuk-button nhsuk-button--secondary nhsuk-u-margin-right-3"
            draggable="false"
            href="/templates/preview-letter-template/template-id"
            id="go-back-button"
            role="button"
          >
            Go back
          </a>
          <button
            aria-disabled="false"
            class="nhsuk-button"
            id="submit-template-button"
            type="submit"
          >
            Submit template
          </button>
        </form>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
